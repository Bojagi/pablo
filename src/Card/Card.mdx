---
name: Card
route: /card
menu: Components
---

import { Props, Playground } from 'docz'
import { Card } from './Card.tsx'
import {PlainCard, CustomStyles} from './Card.stories.tsx'

# Card

Renders a plain card with content inside.

## Import

```js
import { Card } from '@bojagi/pablo/Card';
```

## Examples

### Square
<Playground>
<PlainCard />
</Playground>

### Custom styles
<Playground>
<CustomStyles />
</Playground>

## Props
<Props of={Card} />
