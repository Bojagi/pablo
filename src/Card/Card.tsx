import styled from 'styled-components';
import { layoutInterpolationFn, LayoutBoxProps } from '../Box';
import { getComponentStyle, shadowTransformer } from '../styleHelpers/getComponentStyle';

export const Card = styled.div<LayoutBoxProps>`
  padding: ${getComponentStyle('card.padding')};
  background-color: ${getComponentStyle('card.backgroundColor')};
  color: ${getComponentStyle('card.color')};
  box-shadow: ${getComponentStyle('card.shadow', shadowTransformer)};
  border-radius: ${getComponentStyle('card.borderRadius')};
  ${layoutInterpolationFn}
`;
