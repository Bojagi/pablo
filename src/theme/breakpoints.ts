export type Breakpoint = 'sm' | 'md' | 'lg' | 'xl';

export const breakpointNames: Breakpoint[] = ['sm', 'md', 'lg', 'xl'];

export type BreakpointsArray = [number, number, number, number];

export type Breakpoints = BreakpointsArray &
  Record<Breakpoint, number> & {
    breakpointNames: Breakpoint[];
  };

export function createBreakpoints(
  array: BreakpointsArray,
  usedBreakpointNames: Breakpoint[] = breakpointNames
): Breakpoints {
  const newBreakpoints = ([...array] as unknown) as Breakpoints;
  usedBreakpointNames.forEach((bp, index) => {
    newBreakpoints[bp] = newBreakpoints[index];
  });
  newBreakpoints.breakpointNames = usedBreakpointNames;

  return newBreakpoints as Breakpoints;
}

export const defaultBreakpoints: BreakpointsArray = [600, 1000, 1200, 1920];

export const breakpoints = createBreakpoints(defaultBreakpoints, breakpointNames);
