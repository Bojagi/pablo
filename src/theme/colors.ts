export type ShadedColor = {
  main: string;
  dark: string;
  darkest: string;
  light: string;
  lightest: string;
  contrastText: string;
  contrastTextLight: string;
};

export type CommonColors = {
  white: string;
  black: string;
  whiteContrastText: string;
  blackContrastText: string;
};

export type TextColors = {
  main: string;
  info: string;
};

export type BorderColors = {
  main: string;
  lightest: string;
};

export interface GrayColors {
  '50': string;
  '100': string;
  '200': string;
  '300': string;
  '400': string;
  '500': string;
  '600': string;
  '700': string;
  '800': string;
  '900': string;
}

export type AllColors = ShadedColor & TextColors & BorderColors & CommonColors;

export interface Colors {
  common: CommonColors;
  gray: GrayColors;
  borders: BorderColors;
  brand: ShadedColor;
  positive: ShadedColor;
  negative: ShadedColor;
  neutral: ShadedColor;
  text: TextColors;
}

export const WHITE = '#ffffff';
export const BLACK = '#000000';

export const colors: Colors = {
  common: {
    white: WHITE,
    black: BLACK,
    whiteContrastText: BLACK,
    blackContrastText: WHITE,
  },
  gray: {
    '50': getGrayColor(0.95),
    '100': getGrayColor(0.9),
    '200': getGrayColor(0.8),
    '300': getGrayColor(0.7),
    '400': getGrayColor(0.6),
    '500': getGrayColor(0.5),
    '600': getGrayColor(0.4),
    '700': getGrayColor(0.3),
    '800': getGrayColor(0.2),
    '900': getGrayColor(0.1),
  },
  borders: {
    main: `${BLACK}${getHexByte(0.25)}`,
    lightest: `${BLACK}${getHexByte(0.1)}`,
  },
  text: {
    main: BLACK,
    info: `${BLACK}${getHexByte(0.5)}`,
  },
  brand: {
    lightest: '#EBE7FF',
    light: '#D6CEFF',
    main: '#6A50F2',
    dark: '#4129BD',
    darkest: '#281683',
    contrastText: WHITE,
    contrastTextLight: BLACK,
  },
  positive: {
    lightest: '#DFF6E6',
    light: '#BBECCB',
    main: '#38C765',
    dark: '#2FB85A',
    darkest: '#108B36',
    contrastText: WHITE,
    contrastTextLight: BLACK,
  },
  negative: {
    lightest: '#FFE7E5',
    light: '#FFCFCC',
    main: '#F21D0D',
    dark: '#CA271B',
    darkest: '#871A12',
    contrastText: WHITE,
    contrastTextLight: BLACK,
  },
  neutral: {
    lightest: '#FCF5DB',
    light: '#FFEB9E',
    main: '#FFDD5B',
    dark: '#EDC52A',
    darkest: '#D6AC0B',
    contrastText: WHITE,
    contrastTextLight: BLACK,
  },
};

/**
 * Get hex color byte (e.g. to get transparency)
 * @param value Percentage value (between 0.0 and 1.0)
 */
function getHexByte(value: number) {
  return (Math.round((255 * value) / 16) * 16).toString(16);
}

function getGrayColor(value: number) {
  return `#${getHexByte(value)}${getHexByte(value)}${getHexByte(value)}`;
}
