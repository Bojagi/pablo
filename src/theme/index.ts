export * from './defaultTheme';
export * from './context';
export * from './PabloThemeProvider';
export * from './useComponentStyle';
