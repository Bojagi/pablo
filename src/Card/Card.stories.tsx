import React from 'react';
import { css } from '@emotion/react';
import { Paragraph, Subtitle } from '../Typography';
import { Card } from './Card';

export default {
  title: 'Card',
};

export const PlainCard = () => (
  <Card>
    <Subtitle>This is a card</Subtitle>
    <Paragraph>With some text</Paragraph>
  </Card>
);

export const CustomStyles = () => (
  <Card
    customStyles={{
      root: css`
        background-color: blue;
        color: white;
      `,
    }}
  >
    <Subtitle>This is a card</Subtitle>
    <Paragraph>With some text</Paragraph>
  </Card>
);
