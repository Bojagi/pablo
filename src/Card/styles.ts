import { getSpacing } from '../styleHelpers/getSpacing';
import { themeVars } from '../theme/themeVars';
import { Style } from '../theme/types';
import { BaseStyles } from '../types';

export type CardStyleProperties = 'root';

export interface CardStyles extends BaseStyles<CardStyleProperties> {
  padding: Style;
  borderRadius: Style;
  backgroundColor: Style;
  color: Style;
  shadow: string[];
}

export const cardStyles: CardStyles = {
  padding: getSpacing(2),
  borderRadius: getSpacing(1),
  backgroundColor: themeVars.colors.common.white,
  color: themeVars.colors.common.whiteContrastText,
  shadow: ['0px 1px 2px rgba(0, 0, 0, 0.1)', '0px 4px 10px rgba(0, 0, 0, 0.05)'],
};
