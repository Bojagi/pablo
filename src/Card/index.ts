import './styles';

export * from './Card';
