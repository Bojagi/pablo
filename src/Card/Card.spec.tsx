import { render } from '@testing-library/react';
import React from 'react';
import { PabloThemeProvider } from '../theme';
import { Card } from './Card';

test('Render card', () => {
  const { container } = renderComponent({});
  expect(container).toMatchSnapshot();
});

function renderComponent(props) {
  return render(
    <PabloThemeProvider>
      <Card {...props}>Hello</Card>
    </PabloThemeProvider>
  );
}
