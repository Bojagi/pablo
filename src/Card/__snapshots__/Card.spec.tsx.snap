// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render card 1`] = `
.c0 {
  box-sizing: border-box;
  padding: 16px;
  background-color: #ffffff;
  color: #000000;
  box-shadow: 0px 1px 2px rgba(0,0,0,0.1),0px 4px 10px rgba(0,0,0,0.05);
  border-radius: 8px;
}

<div>
  <div
    class="c0"
  >
    Hello
  </div>
</div>
`;
